### PR TITLE
This PR contains code to add search bar to datatypes

### PR DESCRIPTION
**Pull request by bhatiadheeraj**
This PR is currently under development and will update in the coming days with changes

### CODE REVIEW

**Comment by bhatiadheeraj** [comment]
The following tasks need to be done : 

- [ ] On backspace need to update search results (Solution: create a copy before filtering and use it)
- [ ] Make the filtering code concise. 
- [ ] Format code 
- [ ] Miscellanous

**Comment by bhatiadheeraj** [comment]
@soichih This is good for review now! 😄

File: ui/src/datatypes.vue [hunk 1]
```diff
@@ -1,5 +1,12 @@
 <template>
 <div>
+    <div class="page-header">
+        <div class="search-box onRight">
+            <b-form-input v-model="query" type="text" placeholder="Search Datatypes" @input="change_query_debounce" class="input"/>
+            <icon name="search" class="search-icon" scale="1.5"/>
+            <icon name="times" class="clear-search" scale="1.5" @click="clearQuery()" v-if="query != ''"/>
```
Comment (soichih) [inline ui/src/datatypes.vue:7]: `v-if="query"` should do it?

File: ui/src/datatypes.vue [hunk 2]
```diff
@@ -47,7 +54,7 @@ import pageheader from '@/components/pageheader'
 import contact from '@/components/contact'
 import app from '@/components/app'
 import tags from '@/components/tags'
-
+let query_debounce;
```
Comment (soichih) [inline ui/src/datatypes.vue:57]: let's use cameCasing for all new code. So do `let queryDebounce`. Eventually we will update all variables with old_naming_schema.

File: ui/src/datatypes.vue [hunk 3]
```diff
@@ -103,6 +117,14 @@ export default {
 
         get_datatypes(prefix) {
             if(!this.datatypes) return false;
+            if(this.filtered.length >0){
```
Comment (soichih) [inline ui/src/datatypes.vue:120]: would it be simpler to do just `if(this.filtered.length) {`

File: ui/src/datatypes.vue [hunk 4]
```diff
@@ -103,6 +117,14 @@ export default {
...
+                return this.filtered.filter(d=>{
```
Comment (soichih) [inline ui/src/datatypes.vue:121]: Please do
`this.filtered.filter(d=>d.name.includes(prefix))`

File: ui/src/datatypes.vue [hunk 5]
```diff
@@ -113,6 +135,14 @@ export default {
 
         get_not_datatypes(prefix) {
             if(!this.datatypes) return false;
+            if(this.filtered.length >0){
+                return this.filtered.filter(d=>{
```
Comment (soichih) [inline ui/src/datatypes.vue:139]: please simplify this

**Review by soichih — CHANGES_REQUESTED** [review]
Thank you!

File: ui/src/datatypes.vue [hunk 6]
```diff
@@ -103,32 +115,54 @@ export default {
...
+            //document.location="#"; //clear hash //TODO what is this?
+            sessionStorage.setItem("datatypes.query", this.query);
+            this.load();
+        },
+
+        load(){
+            if(this.query){
+                this.query.split(" ").forEach(q=>{
+                    if(q == "") return;
+                    this.filtered = this.datatypes.filter((datatype) =>
```
Comment (bhatiadheeraj) [inline ui/src/datatypes.vue:153]: will it be better to use `datatype => datatype.name.toLowerCase().includes(q.toLowerCase()))` in one single line or is this okay ?

Comment (bhatiadheeraj) [reply ui/src/datatypes.vue:153]: I am using a filtered array to store the filtered datatypes from all the data types we have and when the query is empty it should show the original datatypes.

Comment (soichih) [reply ui/src/datatypes.vue:153]: Yes, but I think it will only take the last token into an account.. right? Is that how you want it to work?

Comment (bhatiadheeraj) [reply ui/src/datatypes.vue:153]: I want it to filter results with every key press and backspace pressed and if the query becomes empty or null then to make it return the original/unfiltered datatypes

Comment (soichih) [reply ui/src/datatypes.vue:153]: if you enter "mp2rage life eeg" in the query, you will see *all* datatypes that names matches the eeg - but I would think user would expect datatype that has all 3 tokens matching. Are you aware that `this.filtered` is overwritten for each forEach loop - effectively making the search to be performed on the last token?

Comment (bhatiadheeraj) [reply ui/src/datatypes.vue:153]: Ohh now I see! I'll fix that asap

**Pull request merged**